activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -119,7 +119,6 @@ def build(attributes = nil, &block)
       def concat(*records)
         records = records.flatten
         if owner.new_record?
-          load_target
```

Comment:
Removing this line fixes the given issue. The only test that fails if you remove this line is [this one](https://github.com/rails/rails/blob/2612b88abfca6ac80db1923e133d5e7822bcef0b/activerecord/test/cases/associations/join_model_test.rb#L494), and so the other change in this file is an alternative way of keeping that functionality working.